[PLA-9255] Switching to v3 predictor endpoints.

anoto-moniz opened the pull request:
Predictors have been given their own endpoint: /api/v3/predictors. This PR adjusts the SDK to point at that new set of endpoints.

The biggest change is the shape of the data coming back from the endpoints. It still maps onto the existing objects, but requires some changes to handling.

Additionally, some of the endpoint behavior has changed, which the SDK must account for in order to preserve backwards compatibility. Some fields have been removed in the new API, so corresponding fields (or methods) are now deprecated in the SDK. For example, the update endpoint no longer archives or trains. Instead, the SDK makes a separate call for archiving when appropriate, and always makes another call for training. We also deprecate the previous ability to set an archived property on update to trigger archiving, in favor of dedicated archive/restore methods.

The v1 and v2 endpoints remain available unchanged, so users on older versions of the SDK won't experience any problems. Those will eventually go away, but not without thorough communication with the customer teams and lead time to upgrade, as well as tracking their usage on the server.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Maintenance (non-breaking change to assist developers)

### Adherence to team decisions
- [X] I have added tests for 100% coverage
- [X] I have written Numpy-style docstrings for every method and class.
- [X] I have communicated the downstream consequences of the PR to others.
- [X] I have bumped the version in setup.py

anoto-moniz commented on tests/conftest.py:
```diff
@@ -4,6 +4,33 @@
 import pytest
 
 
+def build_predictor_entity(instance, status_name="READY", status_info=[]):
+    user = str(uuid.uuid4())
+    time = '2020-04-23T15:46:26Z'
+    return dict(
+        id=str(uuid.uuid4()),
+        data=dict(
+            name=instance.get("name"),
+            description=instance.get("description"),
+            instance=instance
+        ),
+        metadata=dict(
+            status=dict(
+                name=status_name,
+                info=status_info
+            ),
+            created=dict(
+                user=user,
+                time=time
+            ),
+            updated=dict(
+                user=user,
+                time=time
+            )
+        )
+    )
+    
+
```
Sure, I can convert this to be a Factory.

anoto-moniz commented on tests/resources/test_predictor.py:
```diff
@@ -50,27 +51,80 @@ def test_delete():
+def test_deprecated_archive_via_update(valid_label_fractions_predictor_data):
+    session = FakeSession()
+    pc = PredictorCollection(uuid.uuid4(), session)
+    entity = deepcopy(valid_label_fractions_predictor_data)
+    entity["metadata"]["archived"] = entity["metadata"]["created"]
+    session.set_response(entity)
```
Same as above.

anoto-moniz commented on tests/resources/test_predictor.py:
```diff
@@ -214,22 +248,23 @@ def test_check_update_some():
-    }
-    session.set_response({"updatable": True, "update": response})
+    })
+    session.set_response({"updatable": True, **response})
```
Elaborate.

anoto-moniz replied:
They're equivalent, and I think this is cleaner.

anoto-moniz commented on src/citrine/_rest/predictor_metadata.py:
```diff
@@ -0,0 +1,71 @@
+from citrine._serialization import properties
+
+
+class PredictorMetadata:
```
It's mimicking the structure of the previous classes `AIResource` and `AIMetatdata`, which I'm pretty sure are also always used together, so I assumed there was some reason. I can merge these two into one, sure. But I'm going to keep them separate from `Predictor`, as the combined class will be used by future refactor work we have in the pipeline. If you'd like the name to reflect that now instead of later, just gimme a suggestion and I'd be happy to consider it.

anoto-moniz commented on tests/serialization/__init__.py:
```diff
@@ -1,19 +1,24 @@
     serialized = module.dump()
-    serialized['id'] = data['id']
-    assert serialized == valid_serialization_output(data)
+    if issubclass(moduleClass, Predictor):
+        assert serialized == dict(json["data"].items())
```
A) This method was previously only used by the predictor tests (it's now also referenced by the design space tests).
B) The predictor payload is now completely different than other "modules".

I'll rename this method to make it clear.

kroenlein replied:
Why change how it's treating objects instead of just passing it what the existing function expects?

kroenlein commented on tests/resources/test_predictor.py:
```diff
@@ -50,27 +51,80 @@ def test_delete():
     session = mock.Mock()
     pc = PredictorCollection(uuid.uuid4(), session)
-    session.get_resource.return_value = valid_label_fractions_predictor_data
+    entity = deepcopy(valid_label_fractions_predictor_data)
+    entity["metadata"]["archived"] = entity["metadata"]["created"]
+    session.put_resource.return_value = entity
```
Right, but this, again, seems like a perfect Factory candidate.

anoto-moniz replied:
Oh, your wording and the line you tagged were unclear. You're talking about the `deepcopy`. Because it's data for a predictor with a specific instance type, so not a great fit for a factory.

kroenlein reviewed (dismissed):
LGTM

Wait to merge until https://jenkins.corp.citrine.io/job/nextgen-ITs/27022/ goes green.

anoto-moniz commented:
Rebased and updated the version. The bug I mentioned above has been deployed to development, so I kicked off a new IT run using this branch: https://jenkins.corp.citrine.io/job/nextgen-ITs/27252/

kroenlein reviewed (approved):
Based on previous review, LGTM
